File: go/dictionary.go
package main

import (
	"bufio"
	"log"
	"os"
	"sort"
	"strings"
)

func LoadWords() *[]string {
	file, err := os.Open("../wordlist.txt")
	if err != nil {
		log.Fatalf("failed to open file: %s", err)
	}
	defer file.Close()

	var wordList []string

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		wordList = append(wordList, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		log.Fatalf("failed to read file: %s", err)
	}

	return &wordList
}

type TrieNode struct {
	children    map[rune]*TrieNode
	isEndOfWord bool
}

type Trie struct {
	root *TrieNode
}

func NewTrie() *Trie {
	return &Trie{
		root: &TrieNode{children: make(map[rune]*TrieNode)},
	}
}

func (t *Trie) Insert(word string) {
	node := t.root
	for _, char := range word {
		if _, exists := node.children[char]; !exists {
			node.children[char] = &TrieNode{children: make(map[rune]*TrieNode)}
		}
		node = node.children[char]
	}
	node.isEndOfWord = true
}

func (t *Trie) searchWords(node *TrieNode, letters map[rune]bool, path []rune, result *[]string, centerLetter rune, containsCenterLetter bool) {
	if node.isEndOfWord && containsCenterLetter {
		*result = append(*result, string(path))
	}

	for ch, child := range node.children {
		if letters[ch] {
			if ch == centerLetter {
				t.searchWords(child, letters, append(path, ch), result, centerLetter, true)
			} else {
				t.searchWords(child, letters, append(path, ch), result, centerLetter, containsCenterLetter)
			}
		}
	}
}

func (t *Trie) FindWords(letters string) []string {
	lettersSet := make(map[rune]bool)
	for _, ch := range letters {
		lettersSet[ch] = true
	}
	centerLetter := rune(letters[0])
	result := []string{}
	t.searchWords(t.root, lettersSet, []rune{}, &result, centerLetter, false)
	var pangrams []string
	var others []string
	for _, word := range result {
		if isPangram(word, letters) {
			pangrams = append(pangrams, strings.ToUpper(word))
		} else {
			others = append(others, word)
		}
	}
	sort.Strings(pangrams)
	sort.Strings(others)
	return append(pangrams, others...)
}

func isPangram(word, letters string) bool {
	lettersMap := make(map[rune]bool)
	for _, letter := range letters {
		lettersMap[letter] = false
	}
	for _, letter := range word {
		if _, exists := lettersMap[letter]; exists {
			lettersMap[letter] = true
		}
	}
	for _, found := range lettersMap {
		if !found {
			return false
		}
	}
	return true
}

func MakeDictionary() *Trie {
	dictionary := NewTrie()
	wordList := LoadWords()
	for _, word := range *wordList {
		if len(word) >= 4 {
			dictionary.Insert(word)
		}
	}
	return dictionary
}

File: node/dictionary.js
const fs = require("fs");
const path = require("path");

const wordList = fs
  .readFileSync(path.join(__dirname, "..", "wordlist.txt"), "utf-8")
  .split("\n");

class Trie {
  constructor() {
    this.isWord = false;
    this.children = new Map();
  }

  addWord(word) {
    let curr = this;
    for (const c of word) {
      if (!curr.children.has(c)) curr.children.set(c, new Trie());
      curr = curr.children.get(c);
    }
    curr.isWord = true;
  }

  findWords(letters, words, triePath, centerLetter, containsCenterLetter) {
    if (this.isWord && containsCenterLetter) words.push(triePath.join(""));
    for (const c of letters) {
      if (this.children.has(c)) {
        if (c === centerLetter) {
          this.children
            .get(c)
            .findWords(letters, words, triePath.concat(c), centerLetter, true);
        } else {
          this.children
            .get(c)
            .findWords(
              letters,
              words,
              triePath.concat(c),
              centerLetter,
              containsCenterLetter
            );
        }
      }
    }
  }
}

class Dictionary {
  constructor() {
    this.root = new Trie();
  }

  addWord(word) {
    this.root.addWord(word);
  }

  findWords(letters) {
    const words = [];
    const centerLetter = letters[0];
    this.root.findWords(letters, words, [], centerLetter, false);
    words.sort();
    return words;
  }
}

const dictionary = new Dictionary();
for (const word of wordList) {
  dictionary.addWord(word);
}

module.exports = dictionary;
